---
title: "Multi-Metagenome"
author: "Zerihun B"
date: "April 6, 2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library("vegan")
library("plyr")
library("RColorBrewer")
library("alphahull")
library("ggplot2")
library(reshape2)
```

## Load all data

```{r }
Sample1 <- read.delim("coverage72.csv", header = T, stringsAsFactors = F)
Sample2 <- read.delim("coverage75.csv", header = T, stringsAsFactors = F)
Sample3 <- read.delim("coverage78.csv", header = T, stringsAsFactors = F)

Sample1$Name <- as.numeric(Sample1$Name)
Sample2$Name <- as.numeric(Sample2$Name)
Sample3$Name <- as.numeric(Sample3$Name)

all.samples <- merge(Sample1,Sample2, by="Name", all=T)
all.samples <- merge(all.samples,Sample3, by="Name", all=T)
all.samples <- all.samples[!is.na(all.samples$Name),]
colnames(all.samples) <- c("Name", "sample1.cov","sample1.len", "sample2.cov","sample2.len", "sample3.cov","sample3.len")
all.samples.cov <- subset(all.samples, select = c(Name, sample1.cov, sample2.cov, sample3.cov))
all.samples.len <- subset(all.samples, select = c(Name, sample1.len, sample2.len, sample3.len))
samples.cov.melted <- melt(all.samples.cov, id="Name")
samples.cov.melted <- samples.cov.melted[!is.na(samples.cov.melted$value) & samples.cov.melted$value >=1,] 
samples.len.melted <- melt(all.samples.len, id="Name")
samples.len.melted <- samples.len.melted[!is.na(samples.len.melted$value) & samples.len.melted$value >=100,]
  
# na.values <- all.samples[is.na(all.samples$Average.coverage.x),]

gc <- read.delim("assembly.gc.tab", header = T)
kmer <- read.delim("assembly.kmer.tab", header = T)
ess <- read.table("assembly.orfs.hmm.id.txt", header = F)
ess.tax <- read.delim("assembly.orfs.hmm.blast.tax.tab", header = F)
cons.tax <- read.delim("assembly.tax.consensus.txt", header = T)

colnames(kmer)[1] = "name"
colnames(ess) = c("name", "orf", "hmm.id")
colnames(ess.tax) = c("name", "orf", "phylum")
colnames(cons.tax) = c("name", "phylum", "tax.color", "all.assignments")
```

## Merge all data on scaffolds into a single dataframe d

```{r }
d <- as.data.frame(cbind(Sample1$Name, Sample2$Reference.length, gc$gc, Sample1$Average.coverage, Sample2$Average.coverage, Sample3$Average.coverage), row.names = F)
colnames(d) = c("name", "length", "gc", "Sample1", "Sample2", "Sample3")
d <- merge(d, cons.tax, by = "name", all = T)

# d$phylum <- sub(" group", "", d$phylum)
ff=c("Unknown","Proteobacteria","Proteobacteria","FBC group","Proteobacteria","FBC group","FBC group","Fusobacteria", "Opisthokonta","Proteobacteria","PVC group", "Terrabacteria group")  

#phy <- labels(d$phylum)

levels(d$phylum) <- ff
#d$phylum = factor(d$phylum,labels=ff)

d2 <- subset(d, !is.na(d$length))
d2$tax.color[d2$phylum=="Proteobacteria"] <- 1 
d2$tax.color[d2$phylum=="FBC group"] <- 2 
d2$tax.color[d2$phylum=="Terrabacteria group"] <- 3 
d2$tax.color[d2$phylum=="Fusobacteria"] <- 4 
d2$tax.color[d2$phylum=="PVC group"] <- 5 

```
## Merge all data on essential genes into a single dataframe e.
```{r }
e <- merge(ess, d, by = "name", all.x = T)
e <- merge(e, ess.tax, by = c("name", "orf"), all.x = T)
e <- e[, -c(11, 12)]

```

## Calculate some statistics
```{r}
genome.stats <- matrix(NA, nrow = 0, ncol = 10)
colnames(genome.stats) <- c("total.length", "# scaffolds", "mean.length", "max.length", "gc", "Sample1", "Sample2","Sample3", "tot.ess", "uni.ess")

calc.genome.stats <- function(x, y) matrix(c(sum(x$length), nrow(x), round(mean(x$length), 1), max(x$length), round(sum((x$gc * x$length))/sum(x$length), 1), round(sum((x$Sample1 * x$length))/sum(x$length), 1), round(sum((x$Sample2 * x$length))/sum(x$length), 1), round(sum((x$Sample3 * x$length))/sum(x$length), 1), nrow(y), length(unique(y$hmm.id))), dimnames = list(colnames(genome.stats), ""))
```

## We also define a funtion to extract a subset of scaffolds, called extract.

```{r }
extract <- function(x, a.def, v1, v2) {
    out <- {
    }
    for (i in 1:nrow(x)) {
        if (inahull(a.def, c(v1[i], v2[i]))) 
            out <- rbind(out, x[i, ])
    }
    return(out)
}
```

## Initial calculations
```{r}
calc.genome.stats(d2,e)
ds <- subset(d2, length > 500 & length < 3e4)
es <- subset(e, length > 5000)
ds.melted <- melt(subset(ds,select=Sample1:Sample3))
ds.melted <- subset(ds.melted, value >1)
ds2 <- subset(ds, select = length:Sample3)
ds2 <- melt(ds2, id=c("length","gc"))
```

## Coverage and length plots for the assembly


```{r}
ggplot(samples.cov.melted, aes(x=Name, y=value, fill=variable)) +geom_bar(stat="identity")+xlab("scaffolds")+ylab("coverage")+scale_y_log10()

ggplot(samples.len.melted, aes(x=Name, y=value, fill=variable)) +geom_bar(stat="identity")+xlab("scaffolds")+ylab("Length")+scale_y_log10()

```
## Coverage plot coloured with GC
```{r}

ggplot(ds.melted, aes(x=value, fill=variable)) +geom_histogram(color="blue", position = "dodge")+xlim(0,1000)+scale_y_log10()+xlab("coverage")

ggplot(ds.melted, aes(x=value, color=variable, fill=variable)) +geom_density(alpha=0.1)+xlim(0,100)+xlab("coverage")

ggplot(ds2, aes(x=gc, fill=variable))+geom_histogram(position = "dodge")
plt.gcl <- ggplot(ds2, aes(x=gc, y=length, color=variable))+geom_point(position = "jitter")
ggsave("gcl.pdf", plt.gcl, width = 20, height = 15, units = "cm")

ggplot(ds, aes(x = Sample2, y = Sample3, color = gc, size = length))+xlim(0.2,500)+ylim(0.2,500)+
  xlab("Coverage (S1)")+ylab("Coverage (S2)")+geom_point(alpha = 0.5)+
  scale_size_area(name = "Scaffold length", max_size = 5, breaks=c(5000,10000,15000,20000))+
  scale_colour_gradientn(colours = c("red", "green", "blue"))
```

```{r}
t <- 5
ds$tax.color[is.na(ds$tax.color)] <- 0
for (i in 1:nrow(ds)) {
    if (as.integer(ds$tax.color[i]) < t & as.integer(ds$tax.color[i]) > 0) {
        ds$tax.color[i] <- brewer.pal(5, "Paired")[as.integer(ds$tax.color[i])]
    } else {
        ds$tax.color[i] <- NA
        ds$phylum[i] <- NA
    }
}
pcol2 <- brewer.pal(6, "Paired")
ds$tax.color[ds$phylum=="Proteobacteria"] <- "#ABCEE3"
ds$tax.color[ds$phylum=="FCB group"] <- "#ff0000"
ds$tax.color[ds$phylum=="Opisthokonta"] <- "#33A02C"
ds$tax.color[ds$phylum=="Terrabacteria group"] <- "#B2DF8A"
ds$tax.color[ds$phylum=="Fusobacteria"] <- "#ff00ff"
pcol <- cbind(unique(ds$tax.color)[-1], unique(ds$phylum)[-1])
pcol <- pcol[1:3,1]
#pcol <- pcol[order(pcol[, 2]), 1]
pcol <- append(pcol, c("#ff0000","#33A02C","#B2DF8A"), after=length(pcol))
pcol3 <-c("#ff0000","#33A02C","#B2DF8A","#ff0000","#33A02C","#B2DF8A")
```

## Coverage plot by phylum


```{r}
ggplot(ds, aes(x = Sample1, y = Sample2, size = length, colour = phylum))+
  xlim(0.2,500)+ylim(0.2,500)+xlab("Coverage (S1)")+ylab("Coverage (S2)")+
  geom_point(alpha = 0.1, colour = "black")+geom_point(shape = 1)+
  scale_colour_manual(name = "Phyla", values = pcol3)+
  scale_size_area(name = "Scaffold length", max_size = 10)+
  guides(colour = guide_legend(override.aes = list(alpha = 1, size = 5, shape = 19)))

```
